File: python/01_list_invoices.py
"""
Cynco API Quickstart -- 01: List Invoices

Demonstrates how to authenticate with the Cynco API and fetch
a filtered list of invoices using raw HTTP requests.

Run: python 01_list_invoices.py
"""

import os
import sys

import requests

# Read the API key from the environment.
API_KEY: str = os.environ.get("CYNCO_API_KEY", "")
BASE_URL: str = "https://app.cynco.io/api/v1"

if not API_KEY:
    print("Missing CYNCO_API_KEY environment variable.")
    print("Set it with: export CYNCO_API_KEY='cak_your_api_key_here'")
    sys.exit(1)

# All requests use the same Authorization header.
headers: dict[str, str] = {
    "Authorization": f"Bearer {API_KEY}",
}


def main() -> None:
    """List invoices with filters and display the results."""

    # Fetch sent invoices, sorted by due date (ascending), 10 per page.
    response = requests.get(
        f"{BASE_URL}/invoices",
        headers=headers,
        params={
            "status": "sent",
            "limit": 10,
            "sort": "dueDate",
            "order": "asc",
        },
        timeout=30,
    )

    # Raise an exception for HTTP errors (4xx, 5xx).
    response.raise_for_status()

    body: dict = response.json()

    # The response envelope always has: success, data, pagination, meta.
    pagination = body["pagination"]
    invoices: list[dict] = body["data"]

    print(f"Found {pagination['total']} sent invoices\n")

    for invoice in invoices:
        # Monetary values are in the smallest currency unit (e.g., cents).
        # Divide by 100 to get the display value.
        total = invoice["total"] / 100

        print(
            f"{invoice['invoiceNumber']} | "
            f"{invoice['customerName']} | "
            f"{invoice['currency']} {total:,.2f} | "
            f"Due: {invoice['dueDate']}"
        )

    if pagination["hasMore"]:
        remaining = pagination["total"] - len(invoices)
        print(f"\n... and {remaining} more")


if __name__ == "__main__":
    main()

File: typescript/src/01-list-invoices.ts
/**
 * Cynco API Quickstart -- 01: List Invoices
 *
 * Demonstrates how to authenticate with the Cynco API and fetch
 * a filtered list of invoices using the TypeScript SDK.
 *
 * Run: npm run example:list
 */

import Cynco from "@cynco/sdk";

// Initialize the client with your API key.
// The SDK reads from CYNCO_API_KEY by default, but you can pass it explicitly.
const cynco = new Cynco({
  apiKey: process.env.CYNCO_API_KEY,
});

async function main(): Promise<void> {
  // List invoices with filters.
  // All list methods return a paginated response with { data, pagination }.
  const response = await cynco.invoices.list({
    status: "sent",
    limit: 10,
    sort: "dueDate",
    order: "asc",
  });

  console.log(`Found ${response.pagination.total} sent invoices\n`);

  // Iterate over the returned page of invoices.
  for (const invoice of response.data) {
    // Monetary values are in the smallest currency unit (e.g., cents).
    // Divide by 100 to get the display value.
    const total = (invoice.total / 100).toFixed(2);

    console.log(
      `${invoice.invoiceNumber} | ${invoice.customerName} | ${invoice.currency} ${total} | Due: ${invoice.dueDate}`
    );
  }

  // Check if there are more pages.
  if (response.pagination.hasMore) {
    console.log(
      `\n... and ${response.pagination.total - response.data.length} more`
    );
  }
}

main().catch((error: unknown) => {
  console.error("Failed to list invoices:", error);
  process.exit(1);
});
